File: Python/maximum-xor-product.py
# Time:  O(n)
# Space: O(1)

# greedy
class Solution(object):
    def maximumXorProduct(self, a, b, n):
        """
        :type a: int
        :type b: int
        :type n: int
        :rtype: int
        """
        MOD = 10**9+7
        for i in reversed(xrange(n)):
            base = 1<<i
            if min(a, b)&base == 0:
                a, b = a^base, b^base
        return (a%MOD)*(b%MOD)%MOD

File: C++/maximum-xor-product.cpp
// Time:  O(n)
// Space: O(1)

// greedy
class Solution {
public:
    int maximumXorProduct(long long a, long long b, int n) {
        static const int MOD = 1e9 + 7;

        for (int64_t i = n - 1; i >= 0; --i) {
            const int64_t base = 1ll << i;
            if ((a & base) == (b & base)) {
                a |= base;
                b |= base;
                continue;
            }
            if (a > b) {
                swap(a, b);
            }
            a |= base;
            b &= ~base;
        }
        return (a % MOD) * (b % MOD) % MOD;
    }
};
